### DISCUSSION TITLE
Mute/unmute/solo etc actions are not added directly to undo/redo stack

### DISCUSSION THREAD

**Discussion by irrationalpie7** [audacity/audacity#6476] [Ideas & Feature Requests]
### Bug description

When you undo after muting a track, the last action you took *before* muting will also be undone. The project does appear to revert to the muted/unmuted status that all tracks had before that other action was taken, but I haven't found a way to undo *just* the muting. As far as I can tell this applies to all muting-related actions (muting or unmuting one/several/all tracks, as well as solo or unsolo). Additionally, when you then redo, it will redo that previous action *and* the muting, there's no way to get to the state in between without making manual changes from either the undone or redone state.

This behavior is annoying but manageable for a single track, but when you have many muted and unmuted tracks, quickly hitting "solo" to focus on a single track can be hard to undo.

Additionally to the above behavior, I noticed that in a recently opened project where the *only* action I took was to mute or unmute, undoing did nothing at all; when searching I found #4794 which might be related to that?

### Steps to reproduce

1. Have an unmuted audio track
2. Do some transformation on it (e.g., amplify)
3. Mute it
4. Undo
5. Redo

### Expected behavior

4. The audio track will be unmuted but otherwise remains the same
5. The audio track will be muted

### Actual behavior

4. The audio track is unmuted but the previous transformation (amplify, etc) has also been undone
5. The previous transformation is redone (amplify, etc) & the track gets muted

### Audacity Version

latest stable version (from audacityteam.org/download)

### Operating system

Linux

### Additional context

My version of audacity is 3.5.1; I think I got it from the downloads page but I'm not sure how to check.

I filed this issue as a bug because it involves behavior that seems counterintuitive to me, but it's possible I should have filed it as a feature request because as far as I know this behavior has been around for a while.

**Comment by LWinterberg** [comment]
Mute and solo behave the same way as selections: Their state is stored in the undo stack but they are not distinct items in the undo stack. Doing it this way has some use. Doing it other ways has too: 

* If Solo/Mute are directly undoable, you can get back to a specific mute state. It's also the most straight-forward way to implement it. (Ableton, Reaper)
* If Solo/Mute live outside of the undo stack, you can undo to listen to your undo stack with only certain tracks audible. (Audition, ACID)

A use case in which Audacity's behavior is superior to what you're expecting is if you're listening back deciding whether a change you made fits with an overall project. In that case you might playback the sound while repeatedly muting and unmuting a track to hear the variation. If you then decide you don't like the change, you can just undo the change with one ctrl+z - as opposed to perhaps dozens of undo steps that only mute and unmute the track you needed for previewing.

At the moment, I don't see a strong need to change this behavior, especially as it's consistent with the on/off toggle for realtime effects where toggling back and forth while making a decision is *really* common.

**Reply by irrationalpie7** [reply]
> In that case you might playback the sound while repeatedly muting and unmuting a track to hear the variation

On the other hand, if muting were on the undo/redo stack you could use the undo/redo shortcuts to toggle the track on and off

> especially as it's consistent with the on/off toggle for realtime effects where toggling back and forth while making a decision is really common.

Again, I'd argue making that undo/redo-able would make it _easier_ to toggle. Though it does bring additional considerations, I suppose--does undo/redo ever change whether the real time effects panel is popped out? Because if it's collapsed and the only change was to an effect, then it'll look like the undo/redo isn't doing anything, so if you're undoing a stack of things you might easily lose your place (in general visualization of real time effects--which ones are applied and what their effects are on the audio wave--is an area I feel audacity could still use some work)

**Comment by petersampsonaudacity** [comment]
Note that this is inconsistent with the used of Pan and Gain Sliders and with the use of Envelopes - all of which _**do**_ get added to the History stack.

Therefore, for consistency, I think it makes eminent sense that the Mute and Solo button actions also get added to the History stack - and thus become subject to Undo.

**Reply by LWinterberg** [reply]
envelopes and sliders arguably are part of the mix, whereas mute, solo and the on/off buttons are more like preview functions.

**Reply by irrationalpie7** [reply]
(Editing to make it clearer who I was replying to:)

> whereas mute, solo and the on/off buttons are more like preview functions

If it were possible to undo the solo button, this argument would be stronger. As-is, "muting" is the only concept audacity has for "I don't want to hear this right now," whether that's because I'm not working on that track yet, it's a reference track that is useful to me when I'm working with a certain track or tracks but not the one I'm working on right now, I need to listen extra carefully to a different particular track for a moment, etc. So, while there are some uses of mute that fall under previewing, there are others that don't and there's not necessarily enough information to know which one people want at a particular moment
